### PR TITLE
Replace MockRegexPattern with real regex patterns for quantity and float parsing in YAMLLoader

### PR DESCRIPTION
**Pull request by ftsamis**
This is a fix for #702. For a detailed description check the issue.

### CODE REVIEW

File: tardis/io/util.py [hunk 1]
```diff
@@ -12,6 +13,34 @@
...
+
+    (
+
+        (?=\d|\.\d)              # ...a number (with at least one digit)
+
+        (?P<int>\d*)             # having a (possibly empty) integer part
+
+        (\.(?P<frac>\d*))?       # followed by an optional fractional part
+
+        (E(?P<exp>[-+]?\d+))?    # followed by an optional exponent, or...
```
Comment (yeganer) [inline tardis/io/util.py]: I think both `E` and `e` should be allowed. This pattern wouldn't recognize `1e10`, right?

File: tardis/io/util.py [hunk 2]
```diff
@@ -12,6 +13,34 @@
...
+
+    |
+
+        Inf(inity)?              # ...an infinity, or...
```
Comment (yeganer) [inline tardis/io/util.py]: I think `inf` should be valid, too.

File: tardis/io/util.py [hunk 3]
```diff
@@ -12,6 +13,34 @@
...
+
+    |
+        NaN                      # NaN
```
Comment (yeganer) [inline tardis/io/util.py]: Please add `nan`.

File: tardis/io/util.py (hunk 1 above)
Comment (ftsamis) [inline tardis/io/util.py]: This pattern is compiled with re.IGNORECASE, so both E and e are allowed.

File: tardis/io/util.py (hunk 2 above)
Comment (ftsamis) [inline tardis/io/util.py]: It is, the regex translates to "inf" optionally followed by "inity". If you are talking about the case, we use this pattern with re.IGNORECASE.

**Comment by wkerzendorf** [comment]
@ftsamis you have now programmed into TARDIS things that are built-in to Python. Using the built-in method works - not clear what advantage (other than you learning more advanced regexp) this has?

File: tardis/io/util.py (hunk 3 above)
Comment (ftsamis) [inline tardis/io/util.py]: Regarding lower-case `nan` look at my previous replies.
Python supports and converts both +nan and -nan to just nan. This proves that there is no good reason to support signed nans, but also there is no good reason not to. :P

**Comment by ftsamis** [comment]
@wkerzendorf Regarding the float regex I'd think of it more like fixing/monkey-patching a [broken regex pattern in pyyaml](https://bitbucket.org/xi/pyyaml/src/10c55a624429f28191ef2bc1d9db44684946eb5d/lib/yaml/resolver.py?at=default&fileviewer=file-view-default#resolver.py-177:184). Python has nothing to do with it other than me basing on their pattern for decimals to be as compatible as possible with Python.

`MockRegexPattern` was not a built-in method, instead it was a hack that did cost time, especially in tests. See #702 for a detailed description and some metrics.

File: tardis/io/util.py (hunk 2 above)
Comment (yeganer) [inline tardis/io/util.py]: I forgot that we use `re.IGNORECASE`.

File: tardis/io/util.py (hunk 1 above)
Comment (yeganer) [inline tardis/io/util.py]: Oh right, but I think we should use `e` as afaik most ppl use the lowercase letter.
To be more explicit I'd drop `re.IGNORECASE` and modify the regex slightly.

**Comment by yeganer** [comment]
@wkerzendorf This is a bug in PyYAML and a discussion on [stackoverflow](http://stackoverflow.com/questions/30458977/yaml-loads-5e-6-as-string-and-not-a-number) suggests a fixed regex for floats.
There is also a fork fixing this bug and others because apparently the original project hasn't been updated in a while.

I think this is something worth changing and as we already have a working solution I don't see a reason to not do this.
If you don't trust the above regex, one could use the regex from stackoverflow which matches everything defines as float by the JSON specification and subsequently the YAML spec.

File: tardis/io/util.py (hunk 1 above)
Comment (ftsamis) [inline tardis/io/util.py]: Capital letter `e` for scientific notation is perfectly valid syntax and used widely. Let's not restrict things for no good reason. If we drop the `re.IGNORECASE` we'd have to make this pattern bigger to catch all the variants of `NaN` (`nan`, `NAN`, ...) and `inf`, or restrict syntax and not allow `Inf` (with a capital `i`) for example, which there is no good reason doing that I can think of.

Comment (yeganer) [inline tardis/io/util.py]: I want to match both `e` and `E` but I prefer `e` in the code, that's all :)

**Comment by ftsamis** [comment]
@yeganer How can one not trust a regular expression? It's not like it is a huge one that would take a team of expert parser authors to understand. Did you notice that the pattern on the stackoverflow page you link as more trustworthy does not match `inf` but still requires `.inf`?

And regarding the JSON specification, we do **not** want to follow the JSON specification (if it's the same as the float automaton shown at http://www.json.org/), since it doesn't acknowledge `inf` or `nan` as floats at all. That also means that the pattern you are linking does not strictly follow the JSON specification and it does break the JSON spec in way more than one ways (json does not allow starting with `+`, it doesn't allow starting with `.` and so on).

If anyone has trouble understanding/trusting the regexes I introduce in this PR and wants me to explain them character by character, let me know.

**Pull request closed without merging**